neard/src/cli.rs

```diff
@@ -373,6 +402,11 @@ impl RunCmd {
             actix::System::current().stop();
         });
         sys.run().unwrap();
+        // Wait a few extra seconds to let RocksDB cleanly shut down.
+        // It's unknown how much time the process actually needs.
+        const ROCKS_DB_SLEEP_SECONDS: u64 = 3;
+        info!(target: "neard", "Waiting for {} seconds to let RocksDB shut down cleanly...",ROCKS_DB_SLEEP_SECONDS);
+        std::thread::sleep(Duration::from_secs(ROCKS_DB_SLEEP_SECONDS));
```

Comment:
Yeah, this is quite suboptimal, for two reasons: 

* it penalises every shutdown by three seconds, which feels pretty bad, especially for sandbox  (which might be started and shut-down in tests at all)
* it doesn't guarantee that the DB gets properly closed. If something somewhere runs a "once in 10 seconds" loop or some such, and holds a reference to db, the db won't get closed

The proper way to handle such situations is to:

* rely on RAII -- making sure that all resources are cleaned up when `Drop` is finished
* rely on structured concurrency -- making sure that all outstanding tasks and threads are joined when we exit scopes

I am not sure how to do this properly in actix -- I've spend quite a bit of time reading into actix code at this point, and I am still hopelessly confused as to what the intended coding patterns are there -- it seems that, eg, `SyncArbiter` doesn't have a way to wait for actor termination at all. 

So, I think this might be OK to solve the immediate problems, but this is definitely very messed up and feels like something we should intentionally fix some day.

Reply:
The issue is that the graceful shutdown is performed by another thread. The current hack also breaks a lot of tests, because tests generously use `System.stop()`.

Let me see if it's possible to directly synchronize the `select!` and `ClientActor` drop.

Printing `std::thread::current().id()` I get the following:

```
Feb 22 10:42:04.958  INFO neard: Got DIED, stopping...
Feb 22 10:42:04.959 ERROR near_store::db: RocksDB cancelling all background work: My thread id is ThreadId(4)
Feb 22 10:42:04.959 ERROR neard::cli: Waiting for another thread to close RocksDB. My thread id is ThreadId(1)
Feb 22 10:42:04.959  INFO neard: Waiting for 3 seconds to let RocksDB shut down cleanly...
Feb 22 10:42:04.959 ERROR near_store::db: RocksDB cancelled all background work and dropped: My thread id is ThreadId(4)
```

In some runs RocksDB was dropped by the same thread, but usually by a different one.


The backtrace at the point of dropping RocksDB object is
```
   0: <near_store::db::RocksDB as core::ops::drop::Drop>::drop
   1: core::ptr::drop_in_place<near_store::db::RocksDB>
   2: alloc::sync::Arc<T>::drop_slow
   3: core::ptr::drop_in_place<core::cell::UnsafeCell<near_epoch_manager::EpochManager>>
   4: alloc::sync::Arc<T>::drop_slow
   5: core::ptr::drop_in_place<nearcore::runtime::NightshadeRuntime>
   6: alloc::sync::Arc<T>::drop_slow
   7: core::ptr::drop_in_place<near_chunks::SealsManager>
   8: core::ptr::drop_in_place<near_client::client::Client>
   9: core::ptr::drop_in_place<near_client::client_actor::ClientActor>
  10: core::ptr::drop_in_place<tokio::runtime::task::core::Stage<actix::contextimpl::ContextFut<near_client::client_actor::ClientActor,actix::context::Context<near_client::client_actor::ClientActor>>>>
  11: tokio::runtime::task::harness::Harness<T,S>::shutdown
  12: tokio::macros::scoped_tls::ScopedKey<T>::set
  13: <tokio::task::local::LocalSet as core::ops::drop::Drop>::drop
  14: core::ptr::drop_in_place<actix_rt::runtime::Runtime>
  15: actix_rt::system::SystemRunner::run
  16: neard::cli::NeardCmd::parse_and_run
  17: std::sys_common::backtrace::__rust_begin_short_backtrace
  18: std::rt::lang_start::{{closure}}
  19: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ops/function.rs:259:13
      std::panicking::try::do_call
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:406:40
      std::panicking::try
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:370:19
      std::panic::catch_unwind
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panic.rs:133:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/rt.rs:128:48
      std::panicking::try::do_call
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:406:40
      std::panicking::try
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:370:19
      std::panic::catch_unwind
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panic.rs:133:14
      std::rt::lang_start_internal
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/rt.rs:128:20
  20: main
  21: __libc_start_call_main
  22: __libc_start_main@GLIBC_2.2.5
  23: _start
```